### PR TITLE
Consensus simulation mode and fee estimator

### LINKED ISSUE #2521: Compute nodes should set fees for consensus transactions
After #2504, compute nodes need to be able to specify fees for consensus transactions.

This requires two things:
* Ability to locally estimate how much gas is required for a transaction.
* Determining the gas price (currently this can be just a configuration variable).

### PR DESCRIPTION
**Pull request by kostko**
Fixes #2521 

* Add simulation mode for gas estimation
  
  In simulation mode, transactions are processed as if they were executed normally
  via DeliverTx except some checks are disabled. This is used to estimate gas.

* Add fee estimator and use it in workers

TODO
* [X] Basic simulation mode implementation (probably racy and unsafe).
* [X] Basic fee estimator helper.
* [x] Make simulation mode non-racy and safe.
* [x] Configuration-based price discovery (e.g., price is manually set via config).
* [x] Configurable ceiling for gas fees (to avoid bugs consuming everything).
* [x] Use fee estimator in registration worker.
* [x] E2E tests with parameters that require fees for all transactions.

### CODE REVIEW

File: go/consensus/tendermint/abci/mux.go [hunk 1]
```diff
@@ -646,16 +654,46 @@ func (mux *abciMux) executeTx(ctx *Context, rawTx []byte) error {
...
+
+	return mux.processTx(ctx, tx)
+}
+
+func (mux *abciMux) EstimateGas(caller signature.PublicKey, tx *transaction.Transaction) (transaction.Gas, error) {
+	// As opposed to other transaction dispatch entry points (CheckTx/DeliverTx), this method can
+	// be called in parallel to the consensus layer and to other invocations.
+	//
+	// For simulation mode, time will be filled in by NewContext from last block time.
+	ctx := NewContext(ContextSimulateTx, time.Unix(0, 0), mux.state)
```
Comment (Yawning) [inline go/consensus/tendermint/abci/mux.go]: -> `time.Time{}`

File: go/consensus/tendermint/abci/context.go [hunk 2]
```diff
@@ -41,17 +65,63 @@ type Context struct {
...
+		blockHeight:   appState.blockHeight,
+		logger:        logging.GetLogger("consensus/tendermint/abci").With("mode", mode),
 	}
+
+	switch mode {
+	case ContextInitChain:
+		c.state = appState.deliverTxTree
+	case ContextCheckTx:
+		c.state = appState.checkTxTree
+	case ContextDeliverTx:
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/consensus/tendermint/abci/context.go]: Oh, didn't know that, nice.

File: go/consensus/api/submission.go [hunk 3]
```diff
@@ -43,6 +70,37 @@ func (m *submissionManager) signAndSubmitTx(ctx context.Context, signer signatur
...
+		}
+
+		// Fetch current consensus gas price and compute the fee.
+		var amount *quantity.Quantity
+		amount, err = m.priceDiscovery.GasPrice(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to determine gas price: %w", err)
+		}
+		var gasQuantity quantity.Quantity
+		if err = gasQuantity.FromInt64(int64(gas)); err != nil {
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/consensus/api/submission.go]: Good idea, added.

File: go/common/quantity/quantity.go [hunk 4]
```diff
@@ -79,6 +83,14 @@ func (q *Quantity) FromInt64(n int64) error {
 	return q.FromBigInt(big.NewInt(n))
 }
 
+// FromUint64 converts from an uint64 to a Quantity.
+func (q *Quantity) FromUint64(n uint64) error {
+	if n > math.MaxInt64 {
```
Comment (Yawning) [inline go/common/quantity/quantity.go]: Just handle this? (See: https://golang.org/pkg/math/big/#Int.SetBit)

Comment (kostko) [reply go/common/quantity/quantity.go]: Ah, duh, fixed.

**Pull request merged**